### PR TITLE
[SPARK-15509][ML][SparkR] R MLlib algorithms should support input columns "features" and "label"

### PR DESCRIPTION
**Pull request by keypointt**
https://issues.apache.org/jira/browse/SPARK-15509
## What changes were proposed in this pull request?

Currently in SparkR, when you load a LibSVM dataset using the sqlContext and then pass it to an MLlib algorithm, the ML wrappers will fail since they will try to create a "features" column, which conflicts with the existing "features" column from the LibSVM loader. E.g., using the "mnist" dataset from LibSVM:
`training <- loadDF(sqlContext, ".../mnist", "libsvm")`
`model <- naiveBayes(label ~ features, training)`
This fails with:

```
16/05/24 11:52:41 ERROR RBackendHandler: fit on org.apache.spark.ml.r.NaiveBayesWrapper failed
Error in invokeJava(isStatic = TRUE, className, methodName, ...) : 
  java.lang.IllegalArgumentException: Output column features already exists.
    at org.apache.spark.ml.feature.VectorAssembler.transformSchema(VectorAssembler.scala:120)
    at org.apache.spark.ml.Pipeline$$anonfun$transformSchema$4.apply(Pipeline.scala:179)
    at org.apache.spark.ml.Pipeline$$anonfun$transformSchema$4.apply(Pipeline.scala:179)
    at scala.collection.IndexedSeqOptimized$class.foldl(IndexedSeqOptimized.scala:57)
    at scala.collection.IndexedSeqOptimized$class.foldLeft(IndexedSeqOptimized.scala:66)
    at scala.collection.mutable.ArrayOps$ofRef.foldLeft(ArrayOps.scala:186)
    at org.apache.spark.ml.Pipeline.transformSchema(Pipeline.scala:179)
    at org.apache.spark.ml.PipelineStage.transformSchema(Pipeline.scala:67)
    at org.apache.spark.ml.Pipeline.fit(Pipeline.scala:131)
    at org.apache.spark.ml.feature.RFormula.fit(RFormula.scala:169)
    at org.apache.spark.ml.r.NaiveBayesWrapper$.fit(NaiveBayesWrapper.scala:62)
    at org.apache.spark.ml.r.NaiveBayesWrapper.fit(NaiveBayesWrapper.sca
The same issue appears for the "label" column once you rename the "features" column.
```

The cause is, when using `loadDF()` to generate dataframes, sometimes it’s with default column name `“label”` and `“features”`, and these two name will conflict with default column names `setDefault(labelCol, "label")` and `setDefault(featuresCol, "features")` of `SharedParams.scala`
## How was this patch tested?

Test on my local machine.

### CODE REVIEW

**Comment by keypointt** [comment]
hi @jkbradley do you mind have a look on this one? thanks a lot :)

**Comment by keypointt** [comment]
hi @jkbradley do you mind have a look on this one? thanks a lot :)

**Comment by shivaram** [comment]
@jkbradley Is this important for 2.0 ?

**Comment by shivaram** [comment]
cc @mengxr

**Comment by shivaram** [comment]
@keypointt Is this PR still relevant ?

**Comment by keypointt** [comment]
I'm not sure, I guess this one is skipped and not important anymore?

I can close it if it's not going to be merged

**Comment by shivaram** [comment]
Hmm - the problem still seems to be relevant. @mengxr @junyangq Would one of you be able to look at this ?

**Comment by junyangq** [comment]
@keypointt Can we keep searching (in random or sequential way) until an unused column name has been found?

**Comment by keypointt** [comment]
sure I'll try to scan through all the mllib algorithms

**Comment by junyangq** [comment]
@shivaram Does it sound reasonable to you? Just discussed this with @jkbradley.

**Comment by shivaram** [comment]
Yeah I was going to say that we need to handle cases where `labels_output` is also used. We can just add a numeric suffix maybe ?

**Comment by junyangq** [comment]
Sounds good. That's also what we meant.

File: mllib/src/test/scala/org/apache/spark/ml/r/RWrapperUtilsSuite.scala [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+import org.apache.spark.SparkFunSuite
+import org.apache.spark.ml.feature.{RFormula, RFormulaModel}
+import org.apache.spark.mllib.util.MLlibTestSparkContext
+
+class RWrapperUtilsSuite extends SparkFunSuite with MLlibTestSparkContext {
+
+  test("avoid column name conflicting") {
+    val rFormula = new RFormula().setFormula("label ~ features")
+    val data = spark.read.format("libsvm").load("../data/mllib/sample_libsvm_data.txt")
```
Comment (keypointt) [inline mllib/src/test/scala/org/apache/spark/ml/r/RWrapperUtilsSuite.scala]: Here I used `"../data/"`, I'm not sure if there is a better way to do it, something like `$current_directory/data/mllib/sample_libsvm_data.txt`?

All I found is like this `val data = spark.read.format("libsvm").load("data/mllib/sample_libsvm_data.txt")` https://github.com/apache/spark/blob/master/examples/src/main/scala/org/apache/spark/examples/ml/NaiveBayesExample.scala#L36

File: mllib/src/test/scala/org/apache/spark/ml/feature/RFormulaSuite.scala [hunk 2]
```diff
@@ -54,9 +54,6 @@ class RFormulaSuite extends SparkFunSuite with MLlibTestSparkContext with Defaul
     intercept[IllegalArgumentException] {
       formula.fit(original)
     }
-    intercept[IllegalArgumentException] {
```
Comment (keypointt) [inline mllib/src/test/scala/org/apache/spark/ml/feature/RFormulaSuite.scala]: here is just a duplication of above

File: mllib/src/main/scala/org/apache/spark/ml/r/RWrapperUtils.scala [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+   * would conflict with RFormula default feature and label column names.
+   * Here is to change the column name to avoid "column already exists" error.
+   *
+   * @param rFormula RFormula instance
+   * @param data Input dataset
+   * @return Unit
+   */
+  def checkDataColumns(rFormula: RFormula, data: Dataset[_]): Unit = {
+    if (data.schema.fieldNames.contains(rFormula.getLabelCol)) {
+      logWarning("data containing 'label' column, so change its name to avoid conflict")
```
Comment (felixcheung) [inline mllib/src/main/scala/org/apache/spark/ml/r/RWrapperUtils.scala]: is it possible to include the featurecol name in logging?

File: mllib/src/main/scala/org/apache/spark/ml/r/RWrapperUtils.scala [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+      rFormula.setLabelCol(rFormula.getLabelCol + "_output")
```
Comment (felixcheung) [inline mllib/src/main/scala/org/apache/spark/ml/r/RWrapperUtils.scala]: what if `something_output` is also already in the DataFrame?
should we check for it?
I thought the earlier discussion calls for appending a sequence number, like `something_output1` and incrementing util it is not already there?

Comment (keypointt) [inline mllib/src/main/scala/org/apache/spark/ml/r/RWrapperUtils.scala]: sure I'll add this logic, incrementing until no conflict

**Comment by junyangq** [comment]
LGTM

File: mllib/src/main/scala/org/apache/spark/ml/r/RWrapperUtils.scala [hunk 5]
```diff
@@ -35,13 +35,37 @@ object RWrapperUtils extends Logging {
    */
   def checkDataColumns(rFormula: RFormula, data: Dataset[_]): Unit = {
     if (data.schema.fieldNames.contains(rFormula.getLabelCol)) {
-      logWarning("data containing 'label' column, so change its name to avoid conflict")
-      rFormula.setLabelCol(rFormula.getLabelCol + "_output")
+      val newLabelName = convertToUniqueName(rFormula.getLabelCol, data.schema.fieldNames)
+      logWarning(
+        s"data containing ${rFormula.getLabelCol} column, changing its name to $newLabelName")
```
Comment (felixcheung) [inline mllib/src/main/scala/org/apache/spark/ml/r/RWrapperUtils.scala]: this sounds a bit like we are renaming the existing `label` column?
perhaps just change to `s"data containing ${rFormula.getLabelCol} column, using new name to $newLabelName instead"`?

Comment (keypointt) [inline mllib/src/main/scala/org/apache/spark/ml/r/RWrapperUtils.scala]: sure, I'll change it

File: mllib/src/main/scala/org/apache/spark/ml/r/RWrapperUtils.scala [hunk 6]
```diff
@@ -35,13 +35,37 @@ object RWrapperUtils extends Logging {
...
+      rFormula.setLabelCol(newLabelName)
     }
 
     if (data.schema.fieldNames.contains(rFormula.getFeaturesCol)) {
-      logWarning("data containing 'features' column, so change its name to avoid conflict")
-      rFormula.setFeaturesCol(rFormula.getFeaturesCol + "_output")
+      val newFeaturesName = convertToUniqueName(rFormula.getFeaturesCol, data.schema.fieldNames)
+      logWarning(
+        s"data containing ${rFormula.getFeaturesCol} column, changing its name to $newFeaturesName")
```
Comment (felixcheung) [inline mllib/src/main/scala/org/apache/spark/ml/r/RWrapperUtils.scala]: same here?

File: mllib/src/main/scala/org/apache/spark/ml/r/RWrapperUtils.scala [hunk 7]
```diff
@@ -35,13 +35,37 @@ object RWrapperUtils extends Logging {
    */
   def checkDataColumns(rFormula: RFormula, data: Dataset[_]): Unit = {
     if (data.schema.fieldNames.contains(rFormula.getLabelCol)) {
-      logWarning("data containing 'label' column, so change its name to avoid conflict")
-      rFormula.setLabelCol(rFormula.getLabelCol + "_output")
+      val newLabelName = convertToUniqueName(rFormula.getLabelCol, data.schema.fieldNames)
```
Comment (felixcheung) [inline mllib/src/main/scala/org/apache/spark/ml/r/RWrapperUtils.scala]: nit: i think we end up checking for `label_output` twice, once in `if (data.schema.fieldNames.contains(rFormula.getFeaturesCol))` and second time within `convertToUniqueName`? Perhaps we merge them?

File: mllib/src/main/scala/org/apache/spark/ml/r/RWrapperUtils.scala [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+      logWarning(
+        s"data containing ${rFormula.getLabelCol} column, using new name $newLabelName instead")
+      rFormula.setLabelCol(newLabelName)
+    }
+
+    if (data.schema.fieldNames.contains(rFormula.getFeaturesCol)) {
+      val newFeaturesName = convertToUniqueName(rFormula.getFeaturesCol, data.schema.fieldNames)
+      logWarning(
+        s"data containing ${rFormula.getFeaturesCol} column, using new name $newFeaturesName")
```
Comment (felixcheung) [inline mllib/src/main/scala/org/apache/spark/ml/r/RWrapperUtils.scala]: let's make this consistent with the message above in L40?

File: mllib/src/main/scala/org/apache/spark/ml/r/RWrapperUtils.scala (hunk 7 above)
Comment (keypointt) [inline mllib/src/main/scala/org/apache/spark/ml/r/RWrapperUtils.scala]: ok, it's a better way

Comment (keypointt) [inline mllib/src/main/scala/org/apache/spark/ml/r/RWrapperUtils.scala]: I think in `if (data.schema.fieldNames.contains(rFormula.getFeaturesCol))`, it's checking `label` only

and in `convertToUniqueName ()`,  `_output` will be appended resulting in `label_output `: `var newName = originalName + "_output"`, and then `label_output ` is checked at `while (fieldNames.contains(newName))`

Comment (felixcheung) [inline mllib/src/main/scala/org/apache/spark/ml/r/RWrapperUtils.scala]: fair enough. that makes sense, thanks

**Comment by felixcheung** [comment]
LGTM. @shivaram do you have any other comment?

**Comment by shivaram** [comment]
LGTM - @felixcheung Feel free to merge when its ready

**Comment by felixcheung** [comment]
Merged. I could't change the assignee in the JIRA, somehow - @shivaram could you please do that?

**Pull request closed without merging**